### ISSUE TITLE
Version 1.0.2 and my "biggest" dataset: Forestry testing result

### ISSUE THREAD

**Issue by FabianCretton** [zazuko/cube-creator#659]
First, thank you all for the version 1.0.2, it is a good job and I guess we can all accommodate with what is possible (or not) with that MVP version.

Today I tried working on my biggest file, in the "fab_forestry_normalized" project.
It is originally a single 242MB, which I did normalize into 4 .csv files. The result is 122 MB, which is a first interesting result.
When working on CSV, I might thus propose BAFU to first normalize the big files.  

**Question 1**: does this normalization make sense, or does it make no difference for the Cube Creator ?

**Question 2**: if the normalization makes sense, do you have good tools to propose to normalize csv files, that is:
- extract some specific columns to create a new file
- eliminate all duplicated rows from a file
I did that with a mix of Powershell under windows, notepad++ and Excel. I did succeed, but it was somewhat cumbersome.
Of course, it will not be for lambda users, but to use in some special cases by people who have a little technical knowledge.

After doing the mapping on the normalized files, I did run a transform:
https://gitlab.ldbar.ch/pipelines/cube-creator/-/jobs/9320  

**Question 3 (transform bug ?)**: the transform was long then expected, and the job gave a time-out after 1 hour (the default time-out for jobs if I understand well).  
I say "longer than expected", but I do not know yet the size of the generated triples and thus the time needed for the stardog upload.  
Remarks/questions:  
- It seems that the VSHN "ECONN Reset Error" is no more happening/was corrected, this is good news.  
- I do not understand why a[ first transform](https://gitlab.ldbar.ch/pipelines/cube-creator/-/jobs/9308) (when I had only the main file "Forestry_E_All_Data_(Normalized)_codesOnly.csv " and thus only the observation table without links to other tables) did take about 30 minutes. And why then adding the 3 other small files, and links to those tables, do push the run time to more than 1 hour ?  
- As you can see, the job "failed", but I guess that somehow the job actually succeeded (data finally visible in cube designer). I see that the job failed with a timeout during the "Uploading result to graph" step, and I can guess that all the job was done on the side of the pipeline, and that Stardog was finishing the upload process even after the time-out  
- Why is that that eventhough the job failed with a time-out, the arrow in the "transform" list did become green ? (which can be correct as the job finally succeded)

After a while, I was able to see the results in the Cube Designer, the result including the links to the 3 other tables, which is great.  

Then I wanted to publish
**Question 4 (publish bug ?)**: the publish job took only 3 minutes and stopped with that error: 
"Error: Unexpected "<https://environment.ld.admin.ch/f" on line 2445147."  
Remarks/questions:  
- why did that occure
- why does the publish item on the GUI just stay gray ? (no log accessible from there, no error mentionned).

I hope that those problems are minor problems (whithout willing to minimize your effort), and I guess that if the Forestry use-case can finally pass, it will be a great new for the BAFU.
Regards

**Comment by tpluscode** [comment]
:one:

Normalizing can help. Especially if there are multiple tables generated from a single CSV, you would get a lot of repetition. The transformation has no way to collapse duplicate values and will process all values, even if a million rows produce only a single entry. All of the generated data will be sent to Stardog and it will only happen in the database that duplicate values are collapsed in one. A great waste of time spent in transformation as well as data transfer

:two:

The short answer would be to have the CSVs already produced in normalised form. Having to break up a denormalised file requires similar effort (both in terms of computations but also human work). The pipeline will effectively ingest the entire dataset and the output graph will in fact be normalised. It's just shifting the work from one process to another.

Bottom line, if we cannot change the producers of the CSV, it may not necessarily be make much difference.

:three:

Unfortunately the timeout prevented the complete tracing from saving in our monitoring tool. That said, we have been working on additional changes which should improve the pipeline performance. Suggest that we repeat these tests.

:four:

This is weird. The exact error indicates that the publishing failed on reading the unpublished cube. This and a handful similar errors look like the incoming data would have been cut-off abruptly (notice the incomplete URL which failed inside the parser). I will create a dedicated issue to track this problem
